packages/datadog-instrumentations/src/mocha.js

```diff
@@ -0,0 +1,91 @@
+const { addHook, channel, AsyncResource } = require('./helpers/instrument')
+const shimmer = require('../../datadog-shimmer')
+
+const runTestStartCh = channel('ci:mocha:run-test:start')
+const errorCh = channel('ci:mocha:run-test:error')
+const skipCh = channel('ci:mocha:run-test:skip')
+const runTestEndCh = channel('ci:mocha:run-test:end')
+const runTestEndAsyncCh = channel('ci:mocha:run-test:async-end')
+const runTestsEndCh = channel('ci:mocha:run-tests:end')
+const hookErrorCh = channel('ci:mocha:hook-error')
+const mochaEachCh = channel('ci:mocha:mocha-each')
+
+function isRetry (test) {
+  return test._currentRetry !== undefined && test._currentRetry !== 0
+}
+
+addHook({
+  name: 'mocha',
+  versions: ['>=5.2.0'],
+  file: 'lib/runner.js'
+}, (Runner) => {
+  shimmer.wrap(Runner.prototype, 'runTest', runTest => function () {
+    if (!runTestStartCh.hasSubscribers) {
+      return runTest.apply(this, arguments)
+    }
+
+    if (!isRetry(this.test)) {
+      runTestStartCh.publish(this.test)
+    }
+
+    this.once('test end', AsyncResource.bind(() => {
```

Comment:
Should these handlers also be in the `isRetry` block? Otherwise these could happen multiple times without a corresponding start event.

Reply:
these test end event is *not* called on a retry. On newer versions there's a `retry` event but not on older ones. As I mentioned in https://github.com/DataDog/dd-trace-js/pull/1893/files#r820560009, retries that eventually pass are treated like a single test